### PR TITLE
Make proprietary drivers optional (1/2): pmbootstrap changes

### PR DESCRIPTION
**Pull request by ollieparanoid**
Here are the changes necessary in pmbootstrap to make proprietary software installed onto the device (firmware and userspace drivers) optional (#756). To full close the issue, we need to apply this concept
to all device packages we already have in a follow-up PR.

The `example-example` device package would look like this:
```shell
# Reference: <https://postmarketos.org/devicepkg>
pkgname="device-example-example"
...
subpackages="$pkgname-nonfree-firmware:nonfree_firmware $pkgname-nonfree-userland:nonfree_userland"

build() {
	devicepkg_build $startdir $pkgname
}

package() {
	devicepkg_package $startdir $pkgname
}

nonfree_firmware() {
	depends="linux-firmware-brcm ..."
	pkgdesc="Modem, Wifi, accelerated GPU"
	mkdir "$subpkgdir"
}

nonfree_userland() {
	depends="..."
	pkgdesc="accelerated GPU"
	mkdir "$subpkgdir"
}
```

## Note on building dependencies
This was a bit unexpected for me, but this is how Alpine's `abuild` works: We can override the `depends` in the subpackages, and `abuild` does not try to install them as dependencies before building the package/subpackage (even in `pmbootstrap build --strict` mode where we let `abuild` install the dependencies).
In other words, it is possible to specify dependencies on non-free programs without requiring them at build time to build the device package. So when someone chooses the libre way, the non-free packages will not only be out of the installed system, they will not even get built.
The same goes for the binary repository (where we don't host proprietary packages (see #797): When compiling the device packages, the firmware packages don't get dragged in.

## Changes
* New config file options nonfree_firmware and nonfree_userland, which
  we ask for during "pmbootstrap init" if there are non-free components
  for the selected device.
* We find that out by checking the APKBUILD's subpakages: The non-free
  packages are called $pkgname-nonfree-firmware and
  $pkgname-nonfree-userland.
* During "pmbootstrap init" we also show the pkgdesc of these
  subpackages. Parsing that is implemented in
  pmb.parse._apkbuild.subpkgdesc(). It was not implemented as part of
  the regular APKBUILD parsing, as this would need a change in the
  output format, and it is a lot *less* code if done like in this
  commit.
* pmb/parse/apkbuild.py was renamed to _apkbuild.py, and
  pmb/install/install.py to _install.py: needed to call the function in
  the usual way (e.g. pmb.parse.apkbuild()) but still being able to
  test the individual functions from these files in the test suite.
  We did the same thing for pmb/build/_package.py already.
* Install: New function get_nonfree_packages() returns the non-free
  packages that will be installed, based on the user's choice in
  "pmbootstrap init" and on the subpackages the device has.
* Added test cases and test data (APKBUILDs) for all new code,
  refactored test/test_questions.py to have multiple functions for
  testing the various questions / question types from
  "pmbootstrap init" instead of having it all in one big function.
  This allows to use another aport folder for testing the new
  non-free related questions in init.

### Preparation
```
$ pmbootstrap init
...
Device [qemu-amd64]: nonfree-firmware-and-userland
You are about to do a new device port for 'nonfree-firmware-and-userland'
Continue? (y/n) [y]:
...
$ cp test/testdata/init_questions_device/aports/device/device-nonfree-firmware-and-userland/APKBUILD aports/device/device-nonfree-firmware-and-userland/APKBUILD
$ pmbootstrap checksum device-nonfree-firmware-and-userland
```
### Actual testing
During `install` it should only try to install the non-free packages you've chosen in `init`:
```
$ pmbootstrap init
$ pmbootstrap install
```

### CODE REVIEW

File: pmb/parse/_apkbuild.py [hunk 1]
```diff
@@ -173,3 +173,37 @@ def apkbuild(args, path, check_pkgver=True):
...
+    for line in lines:
+        if in_function:
+            if line.startswith(prefix_pkgdesc):
+                return line[len(prefix_pkgdesc):-2]
+        elif line.startswith(prefix_function):
+            in_function = True
+
+    # Failure
+    if not in_function:
+        raise RuntimeError("Could not find subpackage function, no line starts"
```
Comment (z3ntu) [inline pmb/parse/_apkbuild.py]: okay well you print an error here; it's okay I guess

**Comment by z3ntu** [comment]
Looked through it, nothing bad caught my eye ;)

File: pmb/parse/_apkbuild.py [hunk 2]
```diff
@@ -173,3 +173,37 @@ def apkbuild(args, path, check_pkgver=True):
...
+    :param function: name of the subpackage (e.g. "nonfree_userland")
+    :returns: the subpackage's pkgdesc
+    """
+    # Read all lines
+    with open(path, encoding="utf-8") as handle:
+        lines = handle.readlines()
+
+    # Prefixes
+    prefix_function = function + "() {"
+    prefix_pkgdesc = "\tpkgdesc=\""
```
[the remark this answers is not in the document]
Comment (ollieparanoid) [reply pmb/parse/_apkbuild.py]: We don't check for that yet, but it would be nice if we did that with `shfmt` for example (#110).

**Comment by ollieparanoid** [comment]
@z3ntu: If you have a few more minutes, maybe you could also run the testing steps so this is safe to merge? Thanks a lot for the review!

**Comment by z3ntu** [comment]
The testing steps per se didn't work for me but checking with `apk info | grep nonfree` in the build root showed what I selected was installed or not installed.

**Comment by ollieparanoid** [comment]
Right, the example device APKBUILD does not depend on any kernel, therefore the `/boot` folder does not created, and it fails later in the install process. But what counts are the installed packages, which you have tested properly. Thanks again, shipping! :ship:

**Pull request merged**